### PR TITLE
Add parsing method for ElasticsearchException.generateThrowableXContent()

### PR DESCRIPTION
**Pull request by tlrx**
The output of the `ElasticsearchException.generateThrowableXContent()` method can be parsed back by the `ElasticsearchException.fromXContent()` method.

This commit adds unit tests in the style of the `to-and-from-xcontent` tests we already have for other parsing methods. It also relax the strict parsing of the `ElasticsearchException.fromXContent()` so that it does not throw an exception when custom metadata of type array & objects are parsed.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 1]
```diff
@@ -392,11 +391,92 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+                } else if (STACK_TRACE.equals(currentFieldName)) {
+                    stack = parser.text();
+                } else {
+                    metadata.put(currentFieldName, Collections.singletonList(parser.text()));
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (CAUSED_BY.equals(currentFieldName)) {
+                    cause = fromXContent(parser);
+                } else if (HEADER.equals(currentFieldName)) {
+                    headers.putAll(parser.map());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I think I would parse headers more precisely into a `Map<String, List<String>>`, it shouldn't be a lot of code.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 2]
```diff
@@ -392,11 +391,92 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+                } else {
+                    // Additional metadata added by the metadataToXContent method is ignored
+                    // and skipped, so that the parser does not fail on unknown fields.
+                    parser.skipChildren();
+                }
+            }else if (token == XContentParser.Token.START_ARRAY) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: can you add a whitespace after the bracket please?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 3]
```diff
@@ -392,11 +391,92 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+                // Additional metadata added by the metadataToXContent method is ignored
+                // and skipped, so that the parser does not fail on unknown fields.
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: can we specify that only complex metadata (object and arrays) are not parsed? This will have to be updated though if we move subclasses to use addMetadata whenever possible, later.

Also, I wonder if we can do better. We could parse this unknown info into a map, but then there wouldn't be a place for it in the ElasticsearchException, unless we make metadata `Map<String, Object>`. Let's think about it, it may make sense for the future, not now maybe.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 4]
```diff
@@ -392,11 +391,92 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+
+            if ( token != null && token.isValue()) {
+                if (TYPE.equals(currentFieldName)) {
+                    type = parser.text();
+                } else if (REASON.equals(currentFieldName)) {
+                    reason = parser.text();
+                } else if (STACK_TRACE.equals(currentFieldName)) {
+                    stack = parser.text();
+                } else {
+                    metadata.put(currentFieldName, Collections.singletonList(parser.text()));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: oh actually we skip all arrays in this parser, so I think we support only key-pairs, but we may print out arrays too?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 5]
```diff
@@ -392,11 +391,92 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+            }
+        } while ((token = parser.nextToken()) == XContentParser.Token.FIELD_NAME);
+
+        StringBuilder message = new StringBuilder("Elasticsearch exception [");
+        message.append(TYPE).append('=').append(type).append(", ");
+        message.append(REASON).append('=').append(reason);
+        if (stack != null) {
+            message.append(", ").append(STACK_TRACE).append('=').append(stack);
+        }
+        message.append(']');
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Can we compare messages from a normal ElasticsearchException with and without cause, with ones obtained from a parsed one? I wonder specifically if the `Elasticsearch exception [` prefix is needed etc.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 6]
```diff
@@ -516,6 +518,41 @@ public void testFromXContentWithHeadersAndMetadata() throws IOException {
...
     }
 
+    public void testThrowableToAndFromXContent() throws IOException {
+        final XContent xContent = randomFrom(XContentType.values()).xContent();
+
+        final Tuple<Throwable, ElasticsearchException> exceptions = randomExceptions();
+        final Throwable throwable = exceptions.v1();
+
+        BytesReference throwableBytes = XContentHelper.toXContent((b, p) -> {
+            ElasticsearchException.generateThrowableXContent(b, ToXContent.EMPTY_PARAMS, throwable);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: we may have done this somewhere else but I think for correct-ness we should rather pass p in instead of empty params. I would also love to call them builder and params if possible.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 7]
```diff
@@ -533,4 +570,64 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
...
+                        "reason=Wrong state]", new ElasticsearchException("Elasticsearch exception [type=null_pointer_exception, " +
+                        "reason=Unexpected null value]"));
+                expectedCause.addMetadata("es.line", "9");
+                expectedCause.addMetadata("es.col", "42");
+
+                expected = new ElasticsearchException("Elasticsearch exception [type=exception, reason=Parsing failed]", expectedCause);
+                expected.addHeader("doc_id", "test");
+                break;
+            default:
+                throw new IllegalArgumentException("No randomized exceptions generated for type [" + type + "]");
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: nit: throw UnsupportedOperationException instead?

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 8]
```diff
@@ -516,6 +518,41 @@ public void testFromXContentWithHeadersAndMetadata() throws IOException {
         assertThat(cause.getMetadata("es.index_uuid"), hasItem("_na_"));
     }
 
+    public void testThrowableToAndFromXContent() throws IOException {
+        final XContent xContent = randomFrom(XContentType.values()).xContent();
+
+        final Tuple<Throwable, ElasticsearchException> exceptions = randomExceptions();
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: can we test what happens when one calls `setResources` passing in multiple ids? That's the one place where we need lists to work properly.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 9]
```diff
@@ -533,4 +570,64 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
...
+                ((ElasticsearchException) actual).addHeader("doc_id", "test");
+
+                ElasticsearchException expectedCause = new ElasticsearchException("Elasticsearch exception [type=parsing_exception, " +
+                        "reason=Wrong state]", new ElasticsearchException("Elasticsearch exception [type=null_pointer_exception, " +
+                        "reason=Unexpected null value]"));
+                expectedCause.addMetadata("es.line", "9");
+                expectedCause.addMetadata("es.col", "42");
+
+                expected = new ElasticsearchException("Elasticsearch exception [type=exception, reason=Parsing failed]", expectedCause);
+                expected.addHeader("doc_id", "test");
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: should we rather add header or metadata outside of the switch to any generated exception (randomly)? Also test lists for headers too.

File: core/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsedException = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
+
+        assertNotNull(parsedException);
+        assertThat(parsedException.getHeaderKeys(), hasSize(0));
+        assertThat(parsedException.getMetadataKeys(), hasSize(2));
+        assertThat(parsedException.getMetadata("es.phase"), hasItem(phase));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java]: the es. prefix here is a bit weird, but I don't know how to remove it without breaking bw comp. We have to take into account though that even if the info is available, no users will know how to retrieve it :)

**Review by javanna — CHANGES_REQUESTED** [review]
thanks @tlrx I left some comments

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 4 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Metadata arrays are not supported at all and just skipped - we could decide to support arrays of value but this is out of scope of this pull request I think

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 3 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I agree, we may or may not change metadata to be `Map<String, Object>` but I think we should address this later, out of the scope of this pull request. For now, let's just ensure that the parser does not fail.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 6 above)
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: Sure

File: core/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java (hunk 10 above)
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java]: They can still have a look of the content of getMetadataKeys?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 4 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Maybe we should at least parse List<String> given that we have that in some subclass? I wouldn't extend support for objects etc. here, but just for arrays of strings. that is what the addMetadata supports at the moment. I am not talking about supporting anything that may get printed when overriding metadataToXContent.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 3 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: ok

File: core/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java (hunk 10 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/SearchPhaseExecutionExceptionTests.java]: right they can iterate over them, that is acceptable.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 4 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: You're right - we must support what gets printed and it can print arrays of values. I should have added that but I forgot it on my way... Thanks for raising this.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 1 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Make sense

**Comment by tlrx** [comment]
Thanks @javanna, I missed an important point concerning metadata/header list of values. I updated the code according to your comments, can you have another look please?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 5 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I'm not sure to understand what you mean, but with or without cause you'll end up with something like Elasticsearch exception [type=foo, reason=bar], and with few cycles of toXContent/fromXContent it will end up with `Elasticsearch exception [type=exception, reason=Elasticsearch exception [type=exception, reason=... etc]]` which is not great.

Another bad point is the current "No ElasticsearchException found" in the generateFailureXContent() method which can obfuscate any non ElasticsearchException :(

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 11]
```diff
@@ -392,11 +391,92 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+        return innerFromXContent(parser);
+    }
+
+    private static ElasticsearchException innerFromXContent(XContentParser parser) throws IOException {
+        String type = null, reason = null, stack = null;
+        ElasticsearchException cause = null;
+        Map<String, List<String>> metadata = new HashMap<>();
+        Map<String, Object> headers = new HashMap<>();
+
+        XContentParser.Token token = parser.currentToken();
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Oh, I didn't address this point yet.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 12]
```diff
@@ -392,11 +392,118 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+                    // and skipped, so that the parser does not fail on unknown fields. The parser only
+                    // support metadata key-pairs and metadata arrays of values.
+                    parser.skipChildren();
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                // Parse the array and add each item to the corresponding list of metadata.
+                // Arrays of objects are not supported yet and just ignored and skipped.
+                List<String> values = new ArrayList<>();
+                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                    if (token.isValue() || token == XContentParser.Token.VALUE_NULL) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: out of curiosity, why supporting null?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 5 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I was wondering what the difference is between an exception that is created not going through parsing, and an exception that gets parsed. Wondering about the message specifically.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 13]
```diff
@@ -516,6 +522,42 @@ public void testFromXContentWithHeadersAndMetadata() throws IOException {
...
+        try (XContentParser parser = createParser(xContent, throwableBytes)) {
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsedException = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
+        assertNotNull(parsedException);
+
+        ElasticsearchException expected = exceptions.v2();
+        while (expected != null) {
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: nit: here I would maybe use do while just to make absolutely sure that the first round always runs.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 14]
```diff
@@ -533,4 +575,121 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
...
+                expectedCause.addMetadata("es.line", "9");
+                expectedCause.addMetadata("es.col", "42");
+
+                expected = new ElasticsearchException("Elasticsearch exception [type=exception, reason=Parsing failed]", expectedCause);
+                break;
+            default:
+                throw new UnsupportedOperationException("No randomized exceptions generated for type [" + type + "]");
+        }
+
+        if (actual instanceof ElasticsearchException) {
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: you can probably cast it here and reuse it later

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 15]
```diff
@@ -533,4 +575,121 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
...
+            if (randomBoolean()) {
+                Map<String, List<String>> randomHeaders = new HashMap<>();
+
+                int nbHeaders = randomIntBetween(1, 5);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: super minor nit: in general, when creating maps and lists and knowing the size you could pass it in.

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 16]
```diff
@@ -533,4 +575,121 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
...
     }
+
+    private static Tuple<Throwable, ElasticsearchException> randomExceptions() {
+        Throwable actual;
+        ElasticsearchException expected;
+
+        int type = randomIntBetween(0, 5);
+        switch (type) {
+            case 0:
+                actual = new ClusterBlockException(singleton(DiscoverySettings.NO_MASTER_BLOCK_WRITES));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: oh I see the trouble that I caused you, because not all of them are actually ElasticsearchException. Good test though, much more coverage now!

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 17]
```diff
@@ -392,11 +392,118 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+                } else if (STACK_TRACE.equals(currentFieldName)) {
+                    stack = parser.text();
+                } else {
+                    metadata.put(currentFieldName, Collections.singletonList(parser.text()));
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (CAUSED_BY.equals(currentFieldName)) {
+                    cause = fromXContent(parser);
+                } else if (HEADER.equals(currentFieldName)) {
+                    Map<String, Object> headerMap = parser.map();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: you are still casting here and eventually calling toString on Objects. What I meant is manually parse these headers instead and make the parsing code a little more accurate. That also won't require to go through the map once again once parsed.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 18]
```diff
@@ -392,11 +392,118 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+
+        XContentParser.Token token = parser.currentToken();
+        String currentFieldName = parser.currentName();
+        do {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                token = parser.nextToken();
+            }
+
+            if (token != null && token.isValue()) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Same question as before: how can the token be null? that seems like a bug?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 19]
```diff
@@ -392,11 +392,118 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+        return innerFromXContent(parser);
+    }
+
+    private static ElasticsearchException innerFromXContent(XContentParser parser) throws IOException {
+        String type = null, reason = null, stack = null;
+        ElasticsearchException cause = null;
+        Map<String, List<String>> metadata = new HashMap<>();
+        Map<String, List<String>> headers = new HashMap<>();
+
+        XContentParser.Token token = parser.currentToken();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: same as before: shall we check what the current token is? if it is a field_name we could use a while instead of a do while? that would be a bit more readable I think.

**Review by javanna — APPROVED** [review]
left a few minors, feel free to push once those are addressed.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 12 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: question: what happens with fields that are not strings? do they become strings?

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: It supports null because headers/metadata/resources can contain null values and those null values will be present in the printed xcontent. So I think we have to parse them and it doesn't add extra complexity.



> question: what happens with fields that are not strings? do they become strings?

`parser.textOrNull()` takes care to retrieve a textual representation of the value but we might lose some precision.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 19 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I changed this part a bit. It's tricky to use a `while ((token =  parser.nextToken()) == ...)` like we do in other parsers because we mustn't parse the next token while entering the loop but we still need to do it while leaving the loop (so that the method left the parser in a coherent state, ie right after the last known field has been parsed).

I changed this to use a for loop, hoping it's easier to read now.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 18 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Right, I removed that.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 12 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: you mean they can contain null values because HashMap supports it or because it actually makes sense to have null values in there? I wonder if we should rather validate values and reject null.

Would it be possible to only parse strings and skip the rest for now?

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: > you mean they can contain null values because HashMap supports it or because it actually makes sense to have null values in there?

Neither of the two: since we allow nulls to be added and printed I thought it would be better to also be able to parse them back.



> I wonder if we should rather validate values and reject null.
> 
> Would it be possible to only parse strings and skip the rest for now?

I agree and yes, as long ass we take care of rejecting them in a follow up PR :)

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: 4d430948a8c02b2c2956c5188b0d4d2ac76c4357 updates the code so that only strings are parsed.

**Comment by tlrx** [comment]
@elasticmachine retest this please

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 20]
```diff
@@ -392,11 +392,124 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+
+            if (token.isValue()) {
+                if (TYPE.equals(currentFieldName)) {
+                    type = parser.text();
+                } else if (REASON.equals(currentFieldName)) {
+                    reason = parser.text();
+                } else if (STACK_TRACE.equals(currentFieldName)) {
+                    stack = parser.text();
+                } else {
+                    metadata.put(currentFieldName, Collections.singletonList(parser.text()));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: shall we also skip non strings here?

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 21]
```diff
@@ -533,4 +575,121 @@ private static void assertExceptionAsJson(Exception e, String expectedJson) thro
...
+                        resourceIds = new String[randomIntBetween(1, 3)];
+                        for (int j = 0; j < resourceIds.length; j++) {
+                            resourceIds[j] = frequently() ? randomAsciiOfLength(5) : "";
+                        }
+                    }
+                    actualException.setResources(resourceType, resourceIds);
+                    expected.setResources(resourceType, resourceIds);
+                }
+            }
+        }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: can we test the case where we have numbers which get ignored?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 20 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: If we skip strings here, that would mean that some non-string metadata (like `bytes_wanted` & `bytes_limit` from `CircuitBreakingException` or `line` & `col` from `ParsingException`) won't be parsed back... which is sad I think, but maybe more coherent.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: yes I know, I value consistency here, let's not support things only in specific cases. we will parse those back when/if we will support proper numbers.

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Let's do that then

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 21 above)
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: I added a dedicated test for this.

Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: wonderful, thanks

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 22]
```diff
@@ -392,11 +392,124 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
...
+                        } else {
+                            List<String> values = headers.getOrDefault(currentFieldName, new ArrayList<>());
+                            if (token == XContentParser.Token.VALUE_STRING) {
+                                values.add(parser.text());
+                            } else if (token == XContentParser.Token.START_ARRAY) {
+                                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                                    if (token == XContentParser.Token.VALUE_STRING) {
+                                        values.add(parser.text());
+                                    } else {
+                                        parser.skipChildren();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: is this needed here? I think it does something only when the current token is start array or start object.

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Right, and in other cases it's a noop. But here you could have arrays of objects (see ExportException)

**Comment by tlrx** [comment]
Thanks a lot @javanna - this wasn't fun to review at all.

5.x: e7a6cd406d862464f698d18b6cc367a0f6ee2c15
